Copy bundles to tar

joaopapereira opened the pull request:
Some changes in this PR:

- Introduce the concept of Location in the ImagesLock. We store in these  locations other places where a particular image can be found
- Move the FakeRegistry to the helper package
- Add more functions to FakeRegistry to simulate copy of images and to create Images and Bundles from Random Images
- Refactor the e2e test for bundle copy to use `t.Run` and testify

Fixes #84

joaopapereira committed:
Allow the copy to tar and import of nested bundles

- Moved FakeRegistry from cmd to the test/helpers package
- Introduce the concept of Location in the ImagesLock. We store in these
  locations other places where a particular image can be found

joaopapereira committed:
Refactor copy_from_bundle_image_test.go

- Start using t.Run for better name
- Introduce the logger.Section
- Use testify for assertion

cppforlife commented on go.mod:
```diff
@@ -9,6 +9,7 @@ require (
 	github.com/k14s/difflib v0.0.0-20201103203400-90558b9d63e4
 	github.com/maxbrunsfeld/counterfeiter/v6 v6.3.0
 	github.com/onsi/gomega v1.10.4 // indirect
+	github.com/pkg/errors v0.9.1
```
i think in all cases we have been using fmt.Errorf. any reason why this PR introduces this dependency?

cppforlife commented on pkg/imgpkg/bundle/bundle.go:
```diff
@@ -18,17 +18,27 @@ const (
 
 func NewBundle(ref string, imagesMetadata ctlimg.ImagesMetadata) *Bundle {
-	return &Bundle{plainimg.NewPlainImage(ref, imagesMetadata), imagesMetadata}
+	return &Bundle{plainimg.NewPlainImage(ref, imagesMetadata), imagesMetadata, &singleLayerReader{}}
```
looks like we can use NewBundleWithReader with singleLayerReader?

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -13,8 +13,66 @@ import (
+	return o.recAllImagesLock(map[string]bool{})
+}
+
+func (o *Bundle) recAllImagesLock(processedImgs map[string]bool) (*ImagesLock, error) {
```
recAllImagesLock -> buildAllImagesLock?

cppforlife replied:
when you got some time let's chat what locations would be used for following structure:

```
bundle1
  img4
  bundle2
    img1
  bundle3
    img2
    bundle2
      img1
```

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -13,8 +13,66 @@ import (
+		bundle := NewBundleWithReader(image.Image, o.imgRetriever, o.imagesLockReader)
+		ok, err := bundle.IsBundle()
+		if err != nil {
+			return nil, errors.Wrapf(err, "checking if %s is a bundle", image.Image)
```
Most of our errors all tools start with capital letter.

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -32,7 +90,10 @@ func (o *Bundle) ImagesLockLocalized() (lockconfig.ImagesLock, error) {
 }
 
-func (o *Bundle) readImagesLock(img regv1.Image) (lockconfig.ImagesLock, error) {
+type singleLayerReader struct {
```
nit: {} looks a bit better on the same line

cppforlife commented on pkg/imgpkg/lockconfig/images_lock.go:
```diff
@@ -24,6 +24,7 @@ type ImagesLock struct {
 type ImageRef struct {
 	Image       string            `json:"image,omitempty"`       // This generated yaml, but due to lib we need to use `json`
 	Annotations map[string]string `json:"annotations,omitempty"` // This generated yaml, but due to lib we need to use `json`
+	locations   []string
```
the relationship between Image and locations is not quite clear. i left another comment to chat about where we would be looking for assets. let's talk about this as well.

cppforlife reviewed (commented):
first pass. looked at implementation. after our chat, will take a look at the tests.

cppforlife commented on pkg/imgpkg/lockconfig/images_lock.go:
```diff
@@ -72,7 +82,16 @@ func (c ImagesLock) AsBytes() ([]byte, error) {
 	}
 
-	bs, err := yaml.Marshal(c)
+	// Use the first location instead of the value present in Image
```
not sure why we cannot continue Image here.

joaopapereira replied:
When we need to rewrite the ImagesLock file we want to write the PrimaryLocation that will be Image if the image was not copied, or the bundle repository if the image was copied.

We stopped changing the value o Image on the ImagesLock file

cppforlife replied:
hmm, may be it's better to have a method that can generate ImagesLock with PrimaryLocations (fills in image key).  AsBytes() would just do plain marshal.

joaopapereira replied:
hmm, is there a situation where you want to call `AsBytes` and you do not want `PrimaryLocation` to be the value in the Image field?
If there are cases where we need this I would agree with you and can create a function like `AsBytesWithPrimaryLocation` or something. If not, I think this conversation makes sense in the `AsBytes`.

cppforlife replied:
to me this is more about symmetry between unmarshaling and marshaling. eg if i create ImagesLock with some Images (each ImageRef filled with Image), i shouldnt have to use AddLocation.

joaopapereira replied:
You do not have to because by default the Image field is the `PrimaryLocation` of a particular image until you execute an `AddLocation`.

The call to `PrimaryLocation` here is to abstract the concept of the best location to retrieve a particular image from.

Do you see any use case where you would call `AsBytes` and would not want the Image field to be equal to the `PrimaryLocation` to retrieve an image from?

cppforlife commented on pkg/imgpkg/lockconfig/images_lock.go:
```diff
@@ -93,3 +112,34 @@ func (c ImagesLock) WriteToPath(path string) error {
+		return []string{i.Image}
+	}
+
+	locations := i.locations
```
this actually does not copy locations. you'll have to do `locations := append([]string{}, i.locations...)`.

cppforlife commented on pkg/imgpkg/lockconfig/images_lock.go:
```diff
@@ -93,3 +112,34 @@ func (c ImagesLock) WriteToPath(path string) error {
+	return i.Locations()[0]
+}
+
+func (i *ImageRef) DiscardOtherLocations(viableLocation string) {
```
DiscardOtherLocations -> DiscardLocationsExcept

cppforlife replied:
should we also check if viableLocation same as i.Image?

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -135,18 +136,35 @@ func (o CopyRepoSrc) getSourceImages() (*ctlimgset.UnprocessedImageRefs, error)
+
+			imageRefs, err = imgLock.LocationPrunedImageRefs()
+			if err != nil {
+				return nil, fmt.Errorf("localizing images lock")
```
fmt.Errorf("Pruning image ref locations: %s", err)

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -79,6 +111,23 @@ func (o *ImagesLock) LocalizeImagesLock() (lockconfig.ImagesLock, bool, error) {
+func (o ImagesLock) LocationPrunedImageRefs() ([]lockconfig.ImageRef, error) {
+	var imageRefs []lockconfig.ImageRef
+	for _, imgRef := range o.imagesLock.Images {
+		newImgRef := imgRef
```
locations inside imageRef wont be copied as a separate slice. we should probably add DeepCopy() to ImageRef and use it here.

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -26,6 +26,38 @@ type ImagesLock struct {
+		}
+
+		imgRef.AddLocation(imageInBundleRepo)
+		imgRefs = append(imgRefs, imgRef)
```
instead of appending to a new array you can capture loop index and `o.imagesLock.Images[i] = imgRef`

cppforlife commented on pkg/imgpkg/bundle/images_lock.go:
```diff
@@ -26,6 +26,38 @@ type ImagesLock struct {
+	return o.imagesLock.Images
+}
+
+func (o *ImagesLock) Merge(imgLock ImagesLock) error {
```
since NewImagesLock constructor returns pointer-ed ImagesLock lets accept it here as well.

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -15,6 +15,64 @@ import (
+		return nil, err
+	}
+
+	imgsLock.AddImageRef(lockconfig.ImageRef{
```
i dont think we need this because of https://github.com/vmware-tanzu/carvel-imgpkg/pull/99/files#diff-fae76c4e2e5ddd7698236747d7211e5725cf904a59c4bcd081555471458d80f8R170

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -15,6 +15,64 @@ import (
+	return imgsLock, nil
+}
+
+func (o *Bundle) buildAllImagesLock(processedImgs map[string]bool) (*ImagesLock, error) {
```
if we are not using bool value itself, common convention is to use a struct instead

```
processedImgs map[string]struct{}
```

then use

```
processedImgs[image.Image] = struct{}{}
```

joaopapereira committed:
Enable richgo if it is installed

This provides a easier to read output in the command line when running
tests.
richgo will only be used if it is present in the PATH if not, it uses
the normal go output

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -15,6 +15,61 @@ import (
 )
 
+func (o *Bundle) AllImagesLock() (*ImagesLock, error) {
+	imgsLock, err := o.buildAllImagesLock(map[string]struct{}{})
```
could do a direct return here.
